### PR TITLE
Add API for creating promises based on native wasm functions

### PR DESCRIPTION
**Pull request by sbc100**
This API allows JS promises to be created with a unique integer handle and a native function pointer workload.  The native code can than later resolve or reject the promise using `emscripten_promise_resolve` or `emscripten_promise_reject`.

This change uses the handle allocator from library_wasmfs_opfs.js and moves it into generic code.

Split out from #18376

### CODE REVIEW

File: src/library_eventloop.js [hunk 1]
```diff
@@ -155,6 +155,61 @@ LibraryJSEventLoop = {
...
+  },
+
+  emscripten_promise_resolve__deps: ['$promiseMap'],
+  emscripten_promise_resolve__sig: 'vip',
+  emscripten_promise_resolve: function(id, value) {
+#if RUNTIME_DEBUG
+    err('emscripten_resolve_promise: ' + id);
+#endif
+    promiseMap.get(id).resolve(value);
+    promiseMap.free(id);
```
Comment (tlively) [inline src/library_eventloop.js:200]: How about freeing the id first to get more reuse in case the resolve or reject functions want to allocate more promises?

File: system/include/emscripten/em_js.h [hunk 2]
```diff
@@ -7,6 +7,8 @@
 
 #pragma once
 
+#include <emscripten/em_macros.h>
```
Comment (tlively) [inline system/include/emscripten/em_js.h:10]: Is this related?

**Review by tlively — APPROVED** [review]
Looks great! Do we have a story for how this will work with Asyncify/JSPI? It seems like we would be able to provide an `await` function that could be called from C.

File: src/library.js [hunk 3]
```diff
@@ -3696,6 +3696,36 @@ mergeInto(LibraryManager.library, {
...
+    this.freelist = [];
+    this.get = function(id) {
+#if ASSERTIONS
+      assert(this.allocated[id] !== undefined);
+#endif
+      return this.allocated[id];
+    };
+    this.allocate = function(handle) {
+      let id;
+      if (this.freelist.length > 0) {
```
Comment (sbc100) [inline src/library.js:3712]: @tlively .. for debugging purposes it might be nice to avoid re-using handles?  Or at least reuse them less.. we we can detect the use of a stale handle that happens to point to a newly allocated one.  WDYT?  Perhaps in debug builds we should just never reuse handles?

Comment (sbc100) [reply src/library.js:3712]: Or perhaps we only start re-using handles after there are over N in the free list?   Something like 256 should be enough to ensure we don't reuse handles too eagerly?

File: system/include/emscripten/em_js.h (hunk 2 above)
Comment (sbc100) [reply system/include/emscripten/em_js.h:10]: Yes, its need to include `em_js.h` without including the master `emscripten.h`.. which I do in the new test.

File: src/library_eventloop.js (hunk 1 above)
Comment (sbc100) [reply src/library_eventloop.js:200]: I'm not sure it worth extra line of code (since it would need to stack the promise in a local)... I don't think promise ID reuse is something need to prioritize (do you?) .   In fact, shouldn't we try to avoid it where possible to ease debugging.

**Comment by sbc100** [comment]
> Looks great! Do we have a story for how this will work with Asyncify/JSPI? It seems like we would be able to provide an `await` function that could be called from C.

Yes I guess you could have `emscripten_promise_await(id)` which would only work in Asyncify/JSPI mode ?

File: src/library.js (hunk 3 above)
Comment (tlively) [reply src/library.js:3712]: My gut feeling is that we shouldn't have the behavior of debugging and release builds diverge here. But if we would be ok with never reusing handles in debug mode, maybe we'd be ok with that in release mode, too? I'm not tied to the free list design as long as we have a common utility we can share. Maybe we can have a counter with an assertion that the counter doesn't get over 4 billion or whatever.

**Comment by tlively** [comment]
We could even have `emscripten_promise_then`, `emscripten_promise_all`, `emscripten_promice_race`, etc. etc. and mirror the entire promise API into C for use with JSPI.

File: src/library.js (hunk 3 above)
Comment (sbc100) [reply src/library.js:3712]: For now I will leave this as is.
